nixos/bupstash: init module

Tom-Hubrecht opened the pull request:
## Description of changes

Creates a NixOS module to setup encrypted backups with bupstash.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Tom-Hubrecht committed:
nixos/bupstash: init module

NetaliDev commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,296 @@
+  meta.maintainers = [ lib.maintainers.thubrecht ];
+
+  options.services.bupstash = {
+    enable = mkEnableOption "bupstash backup service.";
```
```suggestion
    enable = mkEnableOption "bupstash backup service";
```

The `.` will be added automatically.

NetaliDev commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,296 @@
+    };
+
+    server = {
+      enable = mkEnableOption "a bupstash server.";
```
```suggestion
      enable = mkEnableOption "a bupstash server";
```

Here also: the `.` will be added automatically.

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,294 @@
+                preStart = ''
+                  mkdir -p send-logs/${name}
+
+                  if [ ! -d '${keysDir}' ]; then
+                    mkdir -p ${keysDir}
+
+                    # Generate new keys
+                    ${bupstash} new-key -o ${keysDir}/master.key
+
+                    ${bupstash} new-sub-key \
+                      -k ${keysDir}/master.key \
+                      -o ${keysDir}/put.key \
+                      --put
+
+                    ${bupstash} new-sub-key \
+                      -k ${keysDir}/master.key \
+                      -o ${keysDir}/list.key \
+                      --list-contents
+
+                  elif [ -f '${keysDir}/master.key' ]; then
+                    echo "[!] The master key for the job '${name}' is present, you might want to store it offline and delete it on the server."
+                  fi
```
I'm strongly opposed to baking in key management, for at least two reasons:
- Secret keys shouldn't be moved across machine boundaries, not even temporarily
- The way you set this up makes it convergent, not declarative

Tom-Hubrecht replied:
It is very reasonable, I removed this mechanism, and added a `key` option to the jobs

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,281 @@
+
+  config = mkMerge [
+    (mkIf (cfg.jobs != {}) {
+      systemd.tmpfiles.rules =
+        mapAttrsToList
+          (name: job: "d /var/lib/bupstash/send-logs/${name} 0700 ${job.user} ${job.group} - -")
+          cfg.jobs;
```
Given that bupstash owns its working directory, can it not create that path itself? It really should, and if it does not, consider creating an upstream issue, so we can eventually remove this hack.

Are you worried about the permissions on that path? Either way, please set `StateDirectoryMode=0700` and consider `UMask = "0077"`.

Tom-Hubrecht replied:
After testing, bupstash does not create this path alone, it can create the file for the repository log, but does not create the parent directories if they do not exist

Tom-Hubrecht replied:
After some more research, it is tied to the call to `sqlite3_open_v2` inside rusqlite, so it might be hard to fix upstream

mweinelt replied:
Just setting `environment.HOME = "/var/lib/bupstash"` should solve this.

> --send-log PATH
>
>    Path to the send log file, defaults to one of the following, in order, provided the appropriate environment variables are set, $BUPSTASH_SEND_LOG, $XDG_CACHE_HOME/.cache/bupstash/bupstash.sendlog or $HOME/.cache/bupstash/bupstash.sendlog.

https://bupstash.io/doc/man/bupstash-put.html

Tom-Hubrecht replied:
From the manpages, 
> The send log only remembers the data previously sent, so for efficient 'put' use, give each backup job a unique send log file.
We would need different `$HOME` for each pair of (`job`,`repository`)

Tom-Hubrecht replied:
When splitting the job into multiple units, the dynamicuser meant that creating state directories for each repo in each job was necessary anyway

mweinelt replied:
Bupstash should really take the repository into account when creating send logs. I think we should file an upstream issue for this.

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,281 @@
+                description = "Group for the bupstash server.";
+              };
+
+              preStash = mkOption {
+                type = lines;
+                default = "";
+                description = "Commands to run before `bupstash put`.";
+              };
+
+              postStash = mkOption {
+                type = lines;
+                default = "";
+                description = "Commands to run after `bupstash put`.";
+              };
```
I think these options drive the lingo adoption too hard, because they are executed before and after `borgstash put`. In the context of a put job, these could just be `preHook` and `postHook`, which signifies that the enduser can hook into the execution at defined times.

```suggestion
              preHook = mkOption {
                type = lines;
                default = "";
                description = "Commands to run before `bupstash put`.";
              };

              postHook = mkOption {
                type = lines;
                default = "";
                description = "Commands to run after `bupstash put`.";
              };
```

mweinelt reviewed (changes requested):
Okay, I skimmed the PR, and there are a bunch of things that bug me.

For one the module should likely adopt the lingo of the upstream project, where possible, to allow easier mapping between the man pages and the interface we provide.

Secondly, I'm not a fan of the whole scripting approach to execution. The put job should only execute exactly one `ExecStart=` command, and that is `bupstash put`. Preconditions that are required for a declarative setup to work need to go into `ExecStartPre=` instead.

Thirdly, we require a NixOS test, to make sure this keep working. This can be delayed until we are sure about the interface design.

Finally, I was looking into bupstash as a backup tool as well and came up with the following module mock-up a year or so ago, that could provide some inspiration:

https://gist.github.com/mweinelt/608a5b7d29a9f6c41d4ff71cc472c214

I also asked @piegamesde to participate in reviewing this PR, because more eyes are important on a module, that many people could potentially rely on.

You can and should expect multiple rounds of reviews on a piece like this.

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,281 @@
+
+                path = [ pkgs.coreutils ] ++ (optional hasSSH pkgs.openssh);
+
+                script = builtins.concatStringsSep "\n" (
+                  [ job.preStash ]
+                  ++ (builtins.map
+                    (
+                      r:
+                      let
+                        repo = if builtins.isString r then r else "ssh://${r.user}@${r.server}/${r.path}";
+                        sendlog = builtins.replaceStrings [ "/" ] [ "_" ] (
+                          if builtins.isString r then r else "ssh-${r.server}-${r.path}"
+                        );
+                      in
+                      ''
+                        ${optionalString (builtins.isAttrs r)
+                          "ssh-keygen -F ${r.server} 2>&1 > /dev/null || ssh-keyscan ${r.server} >> ~/.ssh/known_hosts"}
+                        ${bupstash} init -r ${repo} || echo "[-] Repository '${repo}' already exists for '${name}'."
+                        ${bupstash} put ${optionalString (job.command != null) "--exec"} \
+                          --repository "${repo}" \
+                          --key "${job.key}" \
+                          --send-log "send-logs/${name}/${sendlog}" \
+                          ${builtins.toString job.extraOptions} \
+                          ${
+                            builtins.toString (builtins.attrValues (builtins.mapAttrs (tag: val: "${tag}='${val}'") job.tags))
+                          } \
+                          :: ${if (job.command == null) then builtins.toString job.paths else job.command}
+                      ''
+                    )
+                    job.repositories
+                  )
+                  ++ [ job.postStash ]
+                );
```
With how quirky the CLI of bupstash is, `toGNUCommandLine` is less useful than anticipated.

Tom-Hubrecht replied:
What semantic would be preferable for `preHook` and `postHook` ? When we split the backup into individual units by repositories, we can either run `preHook` before each backup to a repository (which might bring concurrency issues) or once before all repositories (the same applying to `postHook`)

mweinelt replied:
I think the best approach to allow for concurrent execution is to move the hooks into dedicated systemd units. The setup could be as follows:

- pre-hook:
  - `Before=`
    - put-jobs
  - `Type=exec`
  - `RemainAfterExit=true`
- put job(s):
  - `Wants=`
    - pre-hook, post-hook
- post-hook
  - `Type=exec`
  - `After=`
    - put-jobs

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,294 @@
+                  StateDirectory = "bupstash";
+                  WorkingDirectory = "/var/lib/bupstash";
+                  User = job.user;
+                  Group = job.group;
```
I added `CAP_DAC_READ_SEARCH` to the capabilities, so that the service is able to read all files, but not modify them, optionally, a user and group can be specified

Tom-Hubrecht commented:
I changed a lot of things @mweinelt , we now have a target unit for each backup jobs, spawning services for each repository, and the pre-hook and post-hook scripts. `DynamicUser` is set to true, but the jobs have read access to all files.
For the server, I added a `ForceCommand` (that can be disabled) to check that only subcommands of `bupstash` are run, allowing for the auto init to succeed.

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+  cfg = config.services.bupstash;
+
+  bupstash = "${cfg.package}/bin/bupstash";
+  defaultUser = "bupstash-server";
```
The scoping of the variable name here is not ideal, given that this is the default user for the server only.

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+                  )
+                );
+                default = [ ];
+                description = "List of repositories to backup to.";
```
I noticed, by chance, that I can pass a path to point to a local directory, mention that in the description.

mweinelt replied:
I talked with @piegamesde yesterday, and we concluded that the whole repository options is too confining. His upcoming review will likely touch on this subject as well.

Tom-Hubrecht replied:
Described in the repository `kind` option

mweinelt reviewed (changes requested):
Second pass. After these comments are addressed, I think a system test covering a few use cases would be appropriate.

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+    name: job:
+    [
+      {
+        name = "bupstash_prehook-${name}";
```
The idea is to prevent name clashing if someone has the brilliant idea to call their backup `prehook`, but as we only have a target and a timer with the short name it should be fine to use a dash

Tom-Hubrecht replied:
I replaced the name by `bupstash-${name}-prehook`, which allow filtering directly on `bupstash-${name}`

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+              RuntimeDirectory = "bupstash/${name}.${repo}";
+              StateDirectory = "bupstash/${name}.${repo}";
+              WorkingDirectory = "/var/lib/bupstash/${name}.${repo}";
+              DynamicUser = true;
+              User = mkIf (job.user != null) job.user;
+              Group = mkIf (job.group != null) job.group;
```
Can't this create problems when no `bupstash` user really exists and then we have different users with the same same and different uids

mweinelt replied:
With a dynamic user it will only set the pretty name. In case a static user with that name exists, DynamicUser= wil be disabled automatically. At least that is how I understand it.

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+
+          path = [ pkgs.coreutils ];
+
+          script = ''
+            echo "Running ${name} bupstash preHook"
+
+            ${job.preHook}
+          '';
```
The issue is that when the `preHook` (or `postHook` for that matter) is empty, then the unit is considered invalid as no ExecStart not ExecStop is defined, the purpose of the echo was to give something to satisfy systemd. We could set the default value of the hooks to `sleep 0` I guess

Tom-Hubrecht replied:
I did : 
```nix
script = ''
  ${job.preHook}
'';
```
Which removed the echo and still wrote a script (although completely empty in some cases)

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+
+      users.groups = mkIf (cfg.server.group == defaultUser) { ${defaultUser} = { }; };
+
+      services.openssh.extraConfig = mkIf cfg.server.restrictCommands ''
+        Match User ${cfg.server.user}
+          ForceCommand "${pkgs.writeShellScript "only-bupstash" ''
+          set -- $SSH_ORIGINAL_COMMAND
+
+          if [ "$1" != "bupstash" ]; then
+            exit 1;
+          fi
+
+          exec "$@"
+        ''}"
+      '';
```
Yeah, my expectations would be that I have per machine repositories, and I think putting various things into that should work. Untested claim.

Tom-Hubrecht replied:
I tried it, you can indeed have a single repository and multiple keys. Each key will only be able to see its associated content

Tom-Hubrecht replied:
It is now possible using `services.bupstash.repositories.access`

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+            );
+          in
+          nameValuePair "bupstash-${name}-${repo}" {
+            path = [ pkgs.coreutils ] ++ (optional (builtins.isAttrs r) pkgs.openssh);
```
I remove `coreutils`, and put `openssh` always.

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+
+            serviceConfig = {
+              Type = "oneshot";
+              RuntimeDirectory = "bupstash/${name}.${repo}";
```
None actually, I removed it from the units

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+            environment =
+              {
+                BUPSTASH_KEY = job.key;
+                BUPSTASH_SEND_LOG = "/var/lib/bupstash/${name}.${repo}/sendlog";
```
Done

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+        value = {
+          description = "Hook to run before the ${name} bupstash backup.";
+
+          partOf = [ "bupstash-${name}.target" ];
+          requiredBy = [ "bupstash-${name}.target" ];
```
Nothing was preventing it, I used your previous solution (with `RemainAfterExit`) to fix this

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+                );
+                default = [ ];
+                description = "List of repositories to backup to.";
+              };
```
So now we have three ways to configure a repository : 

- `uri`: giving the path (or the `ssh://...` uri) that will be put into `BUPSTASH_REPOSITORY`
- `ssh`, using the attributes already defined and that will put the ssh command into `BUPSTASH_REPOSITORY_COMMAND`
- `command`, allowing an arbitrary command to be set in `BUPSTASH_REPOSITORY_COMMAND`

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+
+            config = {
+              extraOptions = builtins.map (pattern: "--exclude '${pattern}'") config.excludes;
+            };
```
Done

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+              # We assume that a repository is already initialized iff its log exists
+              if [ ! -f "sendlog" ]; then
+                ${bupstash} init
+              fi
```
How about `createRepository`? It more clearly paints the picture, that what will happen, once enabled, is the creation of the repository.

piegamesde commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,501 @@
+          cfg.jobs;
+
+      systemd = {
+        services = builtins.listToAttrs (builtins.concatLists (mapAttrsToList mkJobUnits cfg.jobs));
```
I'd like to use `concatMapAttrs` here. However, this would require changing the function signature of `mkJobUnits` to return an attrset instead of a list of namevaluepairs. As a start, you could wrap the current function in `listToAttrs` and then inline that into the function itself later on. (While you could just wrap the entire `mkJobUnits` body in `listToAttrs`, it would be nicer to have it output attribute sets directly. For the first three static items this should be easy, for the list merging you'd have to find appropriate lib functions)

```suggestion
        services = concatMapAttrs (n: v: lib.listToAttrs (mkJobUnits n v)) cfg.jobs;
```

Also, please prefer using the lib functions over the builtins.

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+      systemd = {
+        services = builtins.listToAttrs (builtins.concatLists (mapAttrsToList mkJobUnits cfg.jobs));
+        timers = builtins.listToAttrs (mapAttrsToList mkJobTimer cfg.jobs);
+        targets = builtins.listToAttrs (mapAttrsToList mkJobTarget cfg.jobs);
```
Indeed

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,378 @@
+  options.services.bupstash = {
+    package = mkPackageOption pkgs "bupstash" { };
+
+    server = {
```
I renamed it to `repositories` and used inspiration from the borgbackup module to setup authorized keys for the repos

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,556 @@
+
+  config = mkMerge [
+    (mkIf (cfg.jobs != { }) {
+      assertions =
```
Add an assertion, that ensures either `paths` or `commands` is configured, never both.

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,556 @@
+                type = str;
+                default = defaultClientUser;
+                description = ''
+                  User for the backup job. The default user '${defaultClientUser}' is not created, so it is only cosmetic.
```
```suggestion
                  User that executes the backup job. Will run as a [dynamic user](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#DynamicUser=) by default.
                  If the backup needs to run under a static UID, declaring a user account with the same name is                   sufficient.
```

I don't think we need to put all the reasoning in there.

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,556 @@
+                description = "List of paths to backup.";
+              };
+
+              excludes = mkOption {
+                type = listOf str;
+                default = [ ];
+                description = "List of patterns to exlude from the backup.";
+              };
```
Add example patterns and a reference to where their format is specified.

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,556 @@
+                description = "List of patterns to exlude from the backup.";
+              };
+
+              command = mkOption {
+                type = nullOr str;
+                default = null;
+                description = "Command used for getting the backup data.";
+              };
```
Add an example command, e.g. `${lib.getBin config.services.postgresql.package}/bin/pgdump mydb` and mention that this is used to back up the stdout of the command.

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,556 @@
+                description = "Command used for getting the backup data.";
+              };
+
+              key = mkOption {
+                type = nullOr path;
+                default = null;
+                description = "Path to the key that will be used to encrypt the data.";
+              };
```
Plausible references:
- https://bupstash.io/doc/guides/Secure%20Offline%20Keys.html
- https://bupstash.io/doc/man/bupstash-new-key.html
- https://bupstash.io/doc/man/bupstash-new-sub-key.html

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,556 @@
+                type = nullOr str;
+                default = null;
+                description = ''
+                  A command to run that must print the key data, can be used instead of `job.key`
+                  to fetch the key from arbitrary locations such as the network or other secret storage.
```
```suggestion
                  Run a command that prints the key to stdout. Mutually exclusive with the `key` option.
                  Useful to retrieve the key from password managers.
```

The structure (contract, limitations, example) is good, the wording of the sentences could be simpler and more on point.

Wondering if we can reference other options through a link?

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,556 @@
+              command = mkOption {
+                type = nullOr str;
+                default = null;
+                description = "Command used for getting the backup data.";
```
```suggestion
                description = ''
                  Run a command that prints the data to stdout.
                '';
```

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,556 @@
+              paths = mkOption {
+                type = listOf path;
+                default = [ ];
+                description = "List of paths to backup.";
```
```suggestion
                description = ''
                  List of file and directory paths to include in the backup.
                '';
```

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,556 @@
+            options = {
+              paths = mkOption {
+                type = listOf path;
+                default = [ ];
```
```suggestion
                default = [ ];
                example = [
                  "/var/lib/"
                  "/var/backup/"
                  "/home/"
                ];
```

mweinelt reviewed (dismissed):
Another round of feedback.

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,556 @@
+      builtins.map
+        (
+          repo:
+          let
+            hash = builtins.hashString "sha1" (if repo.kind == "uri" then repo.uri else repo.command);
+          in
+          {
+            "bupstash-${name}-${repo.kind}-${hash}" = {
```
The easiest way is to change `job.repositories` from a list to an attrset and use the key instead of the hash

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,556 @@
+                      };
+
+                      config = {
+                        command = mkIf (config.kind == "ssh") (
+                          (
+                            s:
+                            "ssh -i ${s.privateKey} ${s.user}@${s.server} ${
+                              optionalString (s.path != null) "bupstash serve ${s.path}"
+                            }"
+                          )
+                            config.ssh
```
Are you refering to the shh force command ? If yes it cannot overwrite it.
If you are talking about the `command` option of the repository, it should fail if the user sets `kind = "ssh"` and `command = "..."`

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,556 @@
+              };
+
+              tags = mkOption {
+                type = attrsOf str;
```
Sadly we need to check that the keys are following the regex, so `strMatching` will not help us here, one option would be to add assertions

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,556 @@
+        script = job.preHook;
+
+        serviceConfig = {
+          # Prevents the unit to run multiple times before the postHook has finished
+          RemainAfterExit = true;
```
When `postHook` is also defined, the `ExecStopPost` will kill the prehook, and if `postHook` is empty, then a dummy posthook unit will be created, whose only action is to stop the prehook.

Tom-Hubrecht committed:
nixos/bupstash: Add a test

piegamesde commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,666 @@
+                          default = false;
+                          description = ''
+                            Automatically creates the repository if not already done.
+                            Will cause the job to fail if activated and the repository already exists.
```
This implies that the user would have to first do a run with the option enabled and then disable it for the subsequent runs. It would be preferable for the option to be idempotent, i.e. that once a repo is created this is ignored.

Tom-Hubrecht replied:
The wording is not precise enough I think, what happens is : 
- If no repo exists and we set createRepo to true, then it will init a repo and remember it
- If the repo already exists before the first run (e.g, a common repo for multiple backups) then when trying to create it it will fail

piegamesde commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,666 @@
+                  submodule (
+                    { config, ... }:
+                    {
+                      options = {
```
I'm sorry for the back and forth, but this section will require a bit of rework. My proposal:

- The `ssh` kind has a low value/complexity ratio in my eyes. I'd rather like to not have it and tell users in the documentation to use `ssh` commands or `ssh://` URIs.
- Then the `kind` option can be removed, by simply using `uri` and `command` exclusively and default null values. (Agreed that this is not great code wise, but I think better than exposing a `type` option to the Nix users, as this is not really that common in NixOS.)
- Not sure about the name of `uri` and `command`, although I see that `path` instead of `uri` would be equally confusing. Maybe it will suffice to document that these map to the `BUPSTASH_REPOSITORY` and `BUPSTASH_REPOSITORY_COMMAND` env variables.
- I'm not sure how important per-repo CLI flags are and would propose to remove `enableSendLog`.
- Maybe rename `writePaths` into `writablePaths` or something like that. Also, what's the use-case for specifying any paths beyond the repository path? Maybe this option could be removed as well.

Tom-Hubrecht replied:
When you want to create a local repository automatically, there is nothing for systemd to bind to, so we need to set `writablePaths` to the parent directory of where the repo will be

Tom-Hubrecht replied:
To remove the `writablePaths` option, we could explicitely disallow setting `createRepository` to true when the repository is a local one, and telling the user to create it first (as it is a local one, running `bupstash-init` is quite trivial), what do you think @mweinelt @piegamesde ?

mweinelt replied:
I'm torn on the `writablePaths` idea, given that issue. This mostly affects backups to local storage, for which I would highly recommend going with explicit user/group settings, because otherwise uid/gid won't be stable, and reasonable permissions on the parent directory.

Please read through the other points that were made and let's discuss the details.

> Not sure about the name of uri and command, although I see that path instead of uri would be equally confusing. Maybe it will suffice to document that these map to the BUPSTASH_REPOSITORY and BUPSTASH_REPOSITORY_COMMAND env variables.

i think this is one of the more important points, and going back to something simpler is the solution:

```nix
# BUPSTASH_REPOSITORY
repositories = mkOption {
  type = listOf str;
  default = [];
};

repositoryCommands = mkOption {
  type = listOf str;
  default = [];
}
```

This makes the fewest assumptions and offers the highest customizability, and requires no further escape hatches.

Tom-Hubrecht replied:
I think the highest customizability would be `attrsOf str`, if not, we would still need to find a way to generate a readable name from a command or a path

Tom-Hubrecht replied:
@mweinelt @piegamesde,
To recap the latest push,
- there is now an attrset `repositories` mapped to `BUPSTASH_REPOSITORY`s and `repositoryCommands` mapped to `BUPSTASH_REPOSITORY_COMMAND`
- I removed the `BindPaths` thing
- The option `createRepositories` is a list of names that must defined in `repositories` or `repositoryCommand`
- The `enableSendlog` option has been removed

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,666 @@
+      };
+
+      "bupstash-${name}-posthook" = mkMerge [
+        (mkIf (job.postHook != "") {
```
You cannot set a script to the empy string for systemd services, as it will not create an `ExecStart` entry, and thens systemd will log an error, so we need to filter out empty hooks anyway

Tom-Hubrecht replied:
The reference for this is https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/lib/systemd-unit-options.nix#L392 so nixpkgs already considers that the empty strings means no script.

Tom-Hubrecht commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,666 @@
+        (
+          key:
+          ''command="${bupstash} serve ${
+            builtins.toString (builtins.map (a: "--allow-${a}") allowed)
```
https://nixos.org/manual/nix/stable/language/builtins.html#builtins-toString says that the elements are joined by spaces

mweinelt replied:
You can fold that into one specialized function call, that is much more common and well understood by most contributors. I think this is a no-brainer.

Tom-Hubrecht replied:
Right, a call to `concatMapStringSep` is what is needed

mweinelt replied:
concatMapString**s**Sep

mweinelt commented on nixos/modules/services/backup/bupstash.nix:
```diff
@@ -0,0 +1,603 @@
+
+          path = [ pkgs.openssh ];
+
+          after = optional (job.preHook != "") "bupstash-${name}-prehook.service";
+          before = optional (job.postHook != "" || job.preHook != "") "bupstash-${name}-posthook.service";
```
```suggestion
          after = optional (job.preHook != "") "bupstash-${name}-prehook.service";
          before = optional (job.postHook != "" || job.preHook != "") "bupstash-${name}-posthook.service";
 
          wants = [
            "bupstash-${name}-prehook.service"
            "bupstash-${name}-posthook.service"
          ];
```

In my tests, which admittedly only targeted a single repository, calling the main service for a repo would not result in hooks being started. Turning around the relationship like this does call the hooks.
